### PR TITLE
byoid: extract and apply the ca fingerprint to policy call

### PR DESCRIPTION
**Pull request by dorileo**
Introduce required code to extract the fingerprint content from openssh cert format. The format documentation can be found here:

https://cvsweb.openbsd.org/src/usr.bin/ssh/PROTOCOL.certkeys?annotate=HEAD

### CODE REVIEW

File: src/pam/pam_oslogin_admin.cc [hunk 1]
```diff
@@ -74,6 +76,18 @@ pam_sm_acct_mgmt(pam_handle_t* pamh, int flags, int argc, const char** argv) {
   url << kMetadataServerUrl << "authorize?email=" << UrlEncode(email)
       << "&policy=adminLogin";
 
+  ssh_auth_info = pam_getenv(pamh, "SSH_AUTH_INFO_0");
+  if (ssh_auth_info != NULL) {
+    size_t fp_len;
+    fp_len = sshca_get_byoid_fingerprint(ssh_auth_info, &fingerprint);
+    if (fp_len < 0) {
+      PAM_SYSLOG(pamh, LOG_ERR, "Failed to parse byoid finterprint");
```
Comment (dorileo) [inline src/pam/pam_oslogin_admin.cc:84]: typo.

File: src/pam/pam_oslogin_login.cc [hunk 2]
```diff
@@ -96,6 +98,18 @@ pam_sm_acct_mgmt(pam_handle_t* pamh, int flags, int argc, const char** argv) {
   url << kMetadataServerUrl << "authorize?email=" << UrlEncode(email)
       << "&policy=login";
 
+  ssh_auth_info = pam_getenv(pamh, "SSH_AUTH_INFO_0");
+  if (ssh_auth_info != NULL) {
+    size_t fp_len;
+    fp_len = sshca_get_byoid_fingerprint(ssh_auth_info, &fingerprint);
+    if (fp_len < 0) {
+      PAM_SYSLOG(pamh, LOG_ERR, "Failed to parse byoid finterprint");
```
Comment (dorileo) [inline src/pam/pam_oslogin_login.cc:106]: typo.

File: test/oslogin_sshca_test.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  "CtENcE11Zl53QHbxbAgx22AAAAIHs6r2AekiTHmmoJMKxAKtKW4qcGq5Ku1+SJ1NLdZh0" \
+  "1AAAAAAAAAAAAAAABAAAAFmZpbmdlcnByaW50QGdvb2dsZS5jb20AAAAaAAAAFmZpbmdl" \
+  "cnByaW50QGdvb2dsZS5jb20AAAAAZNEqzAAAAABmsQ0IAAAAAAAAAEMAAAA7ZmluZ2Vyc" \
+  "HJpbnRAZ29vZ2xlLmNvbT1iODZkYjRjYS0wOWZkLTQyOWUtYjEyMS1hMTI3OTk2MTQwMz" \
+  "IAAAAAAAAAAAAAADMAAAALc3NoLWVkMjU1MTkAAAAgyxEJdP6tUhJY3J/4bgLpzyUojE9" \
+  "6YKzE2t/RAx5l32kAAABTAAAAC3NzaC1lZDI1NTE5AAAAQNEBsSEvp5tVMbKUsjIZ0jEa" \
+  "Yv0T0U/GZoCiLfVm3pcXV3RA8aze+y/pbjv+MOxjmAb4KbRH31/S34UALsyGwQM= fing" \
+  "erprint@google.com"                                                    \
+
+TEST(SSHCATests, TestValidSingleExtCert) {
```
Comment (kelvinhwpoon) [inline test/oslogin_sshca_test.cc:120]: Add a test case for invalid certificates?

File: src/pam/oslogin_sshca.cc [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    {"rsa-sha2-256-cert-v01@openssh.com", sshca_rsa_skip_fields},
+    {"rsa-sha2-512-cert-v01@openssh.com", sshca_rsa_skip_fields},
+    {"ssh-dss-cert-v01@openssh.com", sshca_dsa_skip_fields},
+    {"ssh-ed25519-cert-v01@openssh.com", sshca_ed25519_skip_fields},
+    {"ssh-rsa-cert-v01@openssh.com", sshca_rsa_skip_fields},
+    { },
+};
+
+static int
+sshca_get_string(char **buff, size_t *blen, char **ptr, size_t *len_ptr) {
```
Comment (ericdand) [inline src/pam/oslogin_sshca.cc:45]: Nit: I notice that all over this file, you use the return code to indicate success. Our style guide recommends "Prefer using return values over output parameters: they improve readability, and often provide the same or better performance." cf. https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs

But I understand that in plain C, it is hard to use a string as a return value (returning `char *` requires the receiver to figure out the length of the string all over again), so feel free to ignore this nit. ;)

**Review by ericdand — APPROVED** [review]
/lgtm

**Pull request merged**